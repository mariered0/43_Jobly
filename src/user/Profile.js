import React, { useContext, useEffect, useState } from "react";
import {Redirect} from "react-router-dom";
import userContext from "./UserContext";
import JoblyApi from "../api/api";
import {
  Form,
  FormGroup,
  Label,
  Input,
  Card,
  CardBody,
  Button,
  Alert,
} from "reactstrap";

const Profile = () => {
  const { token, currentUser, setCurrentUser } = useContext(userContext);
  const [errors, setErrors] = useState(null);
  const [saveSuccess, setSaveSuccess] = useState(false);

  useEffect(() => {
    if (!currentUser) return;

    setFormData({
      username: currentUser.username,
      firstName: currentUser.firstName,
      lastName: currentUser.lastName,
      email: currentUser.email,
    });
  }, [currentUser]);

  const [formData, setFormData] = useState(null);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  };

  async function handleSubmit(e) {
    e.preventDefault();
    const data = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      password: currentUser.password,
      email: formData.email,
    };
    try {
      const user = await JoblyApi.updateCurrentUserProf(
        currentUser.username,
        data
      );
      setCurrentUser(user);
      setErrors(null);
      setSaveSuccess(true);
    } catch (err) {
      setErrors((errors) => [err]);
    }
  }

  if (token) {
    return (
      <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
        <h2 className="mb-3">Profile</h2>
        {formData && (
        <Card>
          <CardBody>
            <Form>
              <FormGroup>
                <Label for="username" className="mt-3">
                  Username
                </Label>
                <Input
                  id="username"
                  name="username"
                  type="text"
                  value={formData.username}
                  disabled
                />

                <Label for="firstName" className="mt-3">
                  First name
                </Label>
                <Input
                  id="firstName"
                  name="firstName"
                  type="text"
                  value={formData.firstName}
                  onChange={handleChange}
                />

                <Label for="lastName" className="mt-3">
                  Last name
                </Label>
                <Input
                  id="lastName"
                  name="lastName"
                  type="text"
                  value={formData.lastName}
                  onChange={handleChange}
                />

                <Label for="email" className="mt-3">
                  Email
                </Label>
                <Input
                  id="email"
                  name="email"
                  type="email"
                  value={formData.email}
                  onChange={handleChange}
                />
              </FormGroup>

              {errors && (
                <Alert color="danger">
                  {errors.map((error, idx) => (
                    <p key={idx} className="mb-0">
                      {error}
                    </p>
                  ))}
                </Alert>
              )}

              {saveSuccess && (
                <Alert color="success">Updated successfully.</Alert>
              )}

              <Button type="submit" onClick={handleSubmit}>
                Save
              </Button>
            </Form>
          </CardBody>
        </Card>
        )}
      </div>
    );
  }else {
    return <Redirect to="/" />;
  }
};

export default Profile;
