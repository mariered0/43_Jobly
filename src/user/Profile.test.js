import React from "react";
import { MemoryRouter } from "react-router";
import { render } from "@testing-library/react";
import Profile from "./Profile";
import JoblyApi from "../api/api";

import UserContext from "../user/UserContext";

const token = JoblyApi.token;
const currentUser = jest.fn();

it("renders without crashing", () => {
  render(
    <MemoryRouter>
      <UserContext.Provider value={{ token, currentUser }}>
        <Profile />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ token, currentUser }}>
        <Profile />
      </UserContext.Provider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("shows profile in the form", () => {
  const { getByLabelText } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ token, currentUser }}>
        <Profile />
      </UserContext.Provider>
    </MemoryRouter>
  );
  const username = getByLabelText("Username");
  const firstName = getByLabelText("First name");

  expect(username.value).toBe("");
  expect(firstName.value).toBe("");
});
