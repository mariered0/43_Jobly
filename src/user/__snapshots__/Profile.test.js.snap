// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
  >
    <h2
      class="mb-3"
    >
      Profile
    </h2>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="mt-3 form-label"
              for="username"
            >
              Username
            </label>
            <input
              class="form-control"
              disabled=""
              id="username"
              name="username"
              type="text"
              value=""
            />
            <label
              class="mt-3 form-label"
              for="firstName"
            >
              First name
            </label>
            <input
              class="form-control"
              id="firstName"
              name="firstName"
              type="text"
              value=""
            />
            <label
              class="mt-3 form-label"
              for="lastName"
            >
              Last name
            </label>
            <input
              class="form-control"
              id="lastName"
              name="lastName"
              type="text"
              value=""
            />
            <label
              class="mt-3 form-label"
              for="email"
            >
              Email
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              type="email"
              value=""
            />
          </div>
          <button
            class="btn btn-secondary"
            type="submit"
          >
            Save
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
